File: quant1x/distributed/id/option.go
package id

import "math/bits"

type Option func(*HLC)

func WithClock(now func() int64) Option {
	return func(h *HLC) {
		if now != nil {
			h.now = now
		}
	}
}
func WithSeqSeed(seed uint16) Option   { return func(h *HLC) { h.seed = seed } }
func WithStateFile(path string) Option { return func(h *HLC) { h.store = newFileStateStore(path) } }
func WithStateSyncEvery(every uint32) Option {
	return func(h *HLC) {
		if every > 0 {
			h.syncEvery = every
		}
	}
}
func WithStateStrict() Option { return func(h *HLC) { h.strict = true } }
func WithNodeCount(count uint32) Option {
	return func(h *HLC) {
		if count < 1 {
			count = 1
		}
		h.seqBits = payloadBits - uint8(bits.Len(uint(count)))
		if h.seqBits < 4 {
			panic("distributed/id: node count is too large")
		}
	}
}
func WithSeqBits(seqBits uint8) Option {
	return func(h *HLC) {
		if seqBits < 4 || seqBits > payloadBits-1 {
			panic("distributed/id: invalid seq bits")
		}
		h.seqBits = seqBits
	}
}

File: quant1x/distributed/id/Option.java
package quant1x.distributed.id;

import java.util.function.LongSupplier;

@FunctionalInterface
public interface Option {
    void apply(HLC hlc);

    static Option withClock(LongSupplier now) {
        return hlc -> {
            if (now != null) {
                hlc.setNow(now);
            }
        };
    }

    static Option withSeqSeed(int seed) {
        return hlc -> hlc.setSeed(seed);
    }

    static Option withNodeCount(long count) {
        return hlc -> {
            long nodeCount = Math.max(1L, count);
            int workerBits = Long.SIZE - Long.numberOfLeadingZeros(nodeCount);
            int seqBits = HLC.PAYLOAD_BITS - workerBits;
            if (seqBits < 4) {
                throw new IllegalArgumentException("distributed/id: node count is too large");
            }
            hlc.setSeqBits(seqBits);
        };
    }

    static Option withSeqBits(int bits) {
        return hlc -> {
            if (bits < 4 || bits > HLC.PAYLOAD_BITS - 1) {
                throw new IllegalArgumentException("distributed/id: invalid seq bits");
            }
            hlc.setSeqBits(bits);
        };
    }

    static Option withStateFile(String path) {
        return hlc -> {
            if (path != null && !path.isEmpty()) {
                hlc.setStore(new FileStateStore(path));
            }
        };
    }

    static Option withStateFile(java.nio.file.Path path) {
        return hlc -> {
            if (path != null) {
                hlc.setStore(new FileStateStore(path.toString()));
            }
        };
    }
}
